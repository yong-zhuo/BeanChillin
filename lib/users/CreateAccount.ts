'use server'
import { prisma } from "@/lib/prisma";
import { Prisma } from "@prisma/client";
import bcrypt from 'bcrypt';

export default async function CreateAccount(formData: FormData) {
    const email = formData.get('email') as string;
    const password = formData.get('password') as string;
    const confirmPassword = formData.get('confirm-password') as string;
    const msg = RegisterAuth(email, password, confirmPassword);
    if (msg !== 'ok') return msg;

    try {
        const hashedPass = await bcrypt.hash(password, 10);
        const user = await prisma.user.create({
            data: {
                name: "",
                email,
                password: hashedPass,
                signinType: true
            }
        })
        return 'ok';
    } catch (err) {
        if (err instanceof Prisma.PrismaClientKnownRequestError) {
            if (err.code === 'P2002') {
                return 'Email already exist!';
            }
        }
        return 'Error has occured when creating an account! Please try again!';
    }
}

function RegisterAuth(email: string, password: string, confirmPassword: string): string {
    let msg = '';
    if (!email.endsWith('@gmail.com')) {
        msg = 'Invalid email! email should ends with @gmail.com';
    } else if (password !== confirmPassword) {
        msg = 'Password does not match!';
    } else if (password.length < 8) {
        msg = 'Password is too short! Password should be of length minimum 8';
    } else {
        msg = 'ok';
    }
    return msg;
}
