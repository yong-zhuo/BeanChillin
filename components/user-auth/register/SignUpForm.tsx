"use client";

import { signupFields } from "@/constants/formFields";
import Header from "@/components/common-ui/form/Header";
import Button from "@/components/common-ui/button/Button";
import FormInput from "@/components/common-ui/form/FormInput";
import { fieldState } from "@/types/formFieldsState";
import CreateAccount from "../../../lib/users/CreateAccount";
import { useRouter } from "next/navigation";
import { SubmitHandler, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { signupSchema, type signup } from "@/lib/schemas/signupSchema";
import { signIn } from "next-auth/react";
import { useToast } from "@/components/common-ui/shadcn-ui/toast/use-toast";
import { ToastAction } from "@radix-ui/react-toast";
import { useState } from "react";

//signup fields to be mapped
const fields = signupFields;
let fieldsState: fieldState = {};
fields.forEach((field) => (fieldsState[field.id] = ""));

const SignUpForm = () => {

  const [isLoading, setIsLoading] = useState(false);

  //to use shadcn-ui toast
  const { toast } = useToast();

  //zod validation for signup
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<signup>({ resolver: zodResolver(signupSchema) });

  //reroute user to onboarding page upon successful signup
  const router = useRouter();
  const onSubmit: SubmitHandler<signup> = async (data) => {
    setIsLoading(true);
    try {
      const res = (await CreateAccount(data)) as string;
      if (res !== "ok") {
        throw new Error(res);
      }
      console.log(res);
      await signIn("credentials", {
        email: data.email,
        password: data.password,
        callbackUrl: "/onboard",
      });
    } catch (e) {
      toast({
        variant: "destructive",
        title: "Email has already exist!",
        description:
          "Please register with another email or login with the existing email.",
        action: <ToastAction altText="Dismiss">Dismiss</ToastAction>
      });
      setIsLoading(false);
    }
  };

  return (
    <>
      <Header
        heading="Create an account"
        paragraph="Already have an account? "
        linkName="Login Now"
        linkUrl="/login"
        logo
      />
      <form onSubmit={handleSubmit(onSubmit)} className="mb-4 px-40 pb-8 pt-6">
        {fields.map((field) => (
          <FormInput
            key={field.id}
            labelText={field.labelText}
            name={field.name as keyof signup}
            id={field.id}
            type={field.type}
            placeholder={field.placeholder}
            forRegister={field.forRegister}
            register={register}
            error={errors[field.name as keyof signup]}
          />
        ))}
        <div className="mx-3 px-5 py-1">
          <Button
            text="Create Account"
            action="submit"
            addClass=" text-white bg-pri hover:bg-slate-400"
            state={isLoading}
            height={20}
            width={20}
          />
        </div>
      </form>
    </>
  );
};

export default SignUpForm;
