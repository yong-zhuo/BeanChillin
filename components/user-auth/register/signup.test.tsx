import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import SignUpForm from './SignUpForm';
import "@testing-library/jest-dom/";
import { useRouter } from 'next/navigation';
import CreateAccount from '@/lib/users/CreateAccount';
import { SessionProvider, signIn } from 'next-auth/react';
import { Session } from 'next-auth';
import * as nextAuthReact from 'next-auth/react';
let mockedSession: Session | null = null;
jest.mock('next/navigation', () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
  }),
}));
jest.mock("next-auth/next", () => ({
  getServerSession: jest.fn(() => Promise.resolve(mockedSession)),
}));




describe('SignUpForm', () => {
  test('snapshot matches', () => {
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });
    const { asFragment } = render(
      <SessionProvider
        session={mockedSession}
      >
        <SignUpForm/>
      </SessionProvider>,
    );

    expect(asFragment()).toMatchSnapshot();
    
  });
});




describe('SignUpForm', () => {
  test('does not submit the form when the fields are empty', async () => {
    render(<SignUpForm />);

    // Simulate form submission without providing any input
    fireEvent.click(screen.getByRole('button', { name: 'Create Account' }));

    //error messages expected
    await waitFor(() => {
      expect(screen.getByText('Email is required')).toBeInTheDocument();
      const passwordErrors = screen.getAllByText('At least 8 characters long');
      // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
      expect(passwordErrors).toHaveLength(2);
    });

  });
});