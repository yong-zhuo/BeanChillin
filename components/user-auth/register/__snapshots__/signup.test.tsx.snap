// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm snapshot matches 1`] = `
<DocumentFragment>
  <div
    class="mb-5"
  >
    <div
      class="flex justify-center"
    >
      <div
        class="flex justify-center p-10"
      >
        <img
          alt="Logo"
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/logo/logo.svg"
          style="color: transparent;"
          width="230"
        />
      </div>
    </div>
    <h2
      class="text-center text-nowrap text-2xl md:text-3xl font-extrabold text-gray-700"
    >
      Create an account
    </h2>
    <p
      class="mt-2 text-center text-sm md:text-lg text-gray-500"
    >
      Already have an account?  
      <a
        class="font-medium text-pri hover:underline"
        href="/login"
      >
        Login Now
      </a>
    </p>
  </div>
  <form
    class="m-5 lg:mb-5 lg:px-40 lg:pb-2 lg:pt-6 "
  >
    <div
      class="flex justify-between"
    >
      <label
        class="text-md text-black-700 mb-2 block font-bold sm:text-base md:text-sm"
        for="email"
      >
        Email 
        <span
          class="text-pri"
        >
          *
        </span>
      </label>
    </div>
    <div
      class="mb-7"
    >
      <label
        class="hidden"
        for="email"
      >
        Email
      </label>
      <input
        class="shadow rounded-md appearance-none relative block w-full px-3 py-2 border placeholder-gray-400 text-gray-900 focus:outline-none focus:ring-pri focus:border-pri focus:z-10 sm:text-sm "
        data-testid="email"
        name="email"
        placeholder="Enter email address"
        type="email"
      />
    </div>
    <div
      class="flex justify-between"
    >
      <label
        class="text-md text-black-700 mb-2 block font-bold sm:text-base md:text-sm"
        for="password"
      >
        Password 
        <span
          class="text-pri"
        >
          *
        </span>
      </label>
    </div>
    <div
      class="mb-7"
    >
      <label
        class="hidden"
        for="password"
      >
        Password
      </label>
      <input
        class="shadow rounded-md appearance-none relative block w-full px-3 py-2 border placeholder-gray-400 text-gray-900 focus:outline-none focus:ring-pri focus:border-pri focus:z-10 sm:text-sm "
        data-testid="password"
        name="password"
        placeholder="Enter password"
        type="password"
      />
    </div>
    <div
      class="flex justify-between"
    >
      <label
        class="text-md text-black-700 mb-2 block font-bold sm:text-base md:text-sm"
        for="confirm"
      >
        Confirm Password 
        <span
          class="text-pri"
        >
          *
        </span>
      </label>
    </div>
    <div
      class="mb-7"
    >
      <label
        class="hidden"
        for="confirm"
      >
        Confirm Password
      </label>
      <input
        class="shadow rounded-md appearance-none relative block w-full px-3 py-2 border placeholder-gray-400 text-gray-900 focus:outline-none focus:ring-pri focus:border-pri focus:z-10 sm:text-sm "
        data-testid="confirm"
        name="confirm"
        placeholder="Confirm password"
        type="password"
      />
    </div>
    <div
      class="mx-3 px-5 py-1"
    >
      <button
        class="rounded-md w-full group relative flex justify-center border border-transparent text-sm font-medium py-2 px-4 mt-9 text-white bg-pri hover:bg-slate-400 shadow transition ease-in-out duration-300 delay-50 hover:-translate-y-1 hover:scale-110"
        data-testid="Create Account"
        type="submit"
      >
        <div
          class="flex flex-row items-center gap-1 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          Create Account
        </div>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
